New: Site & User options API [ED-5573]

matipojo opened the pull request.

matipojo committed:
New: Site & User options API [ED-5573]

matipojo committed:
Fix PHP Unit test

StyleShit commented on core/options/manager.php:
```diff
@@ -0,0 +1,56 @@
+	/**
+	 * @param Option_Base|string $classname
+	 */
+	public function register( $classname ) {
```
Add an `unregister()` method so it'll be consistent with other managers

matipojo replied:
In this case we won't the ability to unregister

StyleShit commented on core/options/manager.php:
```diff
@@ -0,0 +1,56 @@
+	public function __construct() {
+		parent::__construct();
+
+		do_action( 'elementor/options/register' );
```
```suggestion
		do_action( 'elementor/options/register', $this );
```

StyleShit commented on core/options/manager.php:
```diff
@@ -0,0 +1,56 @@
+	public function register( $classname ) {
+		$this->registered[ $classname::get_key() ] = $classname::get_key();
+
+		if ( method_exists( $classname, 'on_change' ) ) {
```
Looks like `on_change` always exists, since it's defined in the base class

StyleShit commented on core/options/manager.php:
```diff
@@ -0,0 +1,56 @@
+	 */
+	public function get( $key ) {
+		if ( ! isset( $this->registered[ $key ] ) ) {
+			wp_die( esc_attr( $key ) . ' is not a valid option.' );
```
Why not exception?

StyleShit commented on core/options/manager.php:
```diff
@@ -0,0 +1,56 @@
+			wp_die( esc_attr( $key ) . ' is not a valid option.' );
+		}
+
+		return $this->registered[ $key ];
```
Maybe return all when a `$key` is not set?

matipojo replied:
I added a `get_all` method

StyleShit commented on core/options/user-option.php:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	public static function get() {
+		$value = get_user_option( static::get_full_key(), static::$user_id );
+		if ( false === $value ) {
+			$value = static::get_default();
+		}
+		return $value;
```
```suggestion
		$value = get_user_option( static::get_full_key(), static::$user_id );
		
		return ( false === $value ) ? static::get_default() : $value;
```

Or at least add some space:
```suggestion
		$value = get_user_option( static::get_full_key(), static::$user_id );
		
		if ( false === $value ) {
			$value = static::get_default();
		}
		
		return $value;
```

StyleShit commented on includes/beta-testers.php:
```diff
@@ -19,6 +21,9 @@ class Beta_Testers {
 	const NEWSLETTER_PRIVACY_URL = 'https://go.elementor.com/beta-testers-newsletter-privacy';
 
+	/**
+	 * @deprcated Use User_Beta_Tester::SIGNUP
```
```suggestion
	 * @deprecated 3.5.0 Use `User_Beta_Tester::SIGNUP` instead.
```

Should also be added to the deprecations document.

StyleShit commented on includes/tracker.php:
```diff
@@ -60,7 +63,7 @@ public static function init() {
 	 */
 	public static function check_for_settings_optin( $new_value ) {
-		$old_value = get_option( 'elementor_allow_tracking', 'no' );
+		$old_value = Site_Usage_Opt_In::get();
 		if ( $old_value !== $new_value && 'yes' === $new_value ) {
```
```suggestion
		$old_value = Site_Usage_Opt_In::get();
		
		if ( $old_value !== $new_value && 'yes' === $new_value ) {
```

StyleShit commented on includes/tracker.php:
```diff
@@ -162,9 +165,11 @@ public static function send_tracking_data( $override = false ) {
 	 * @since 1.0.0
 	 * @access public
 	 * @static
+	 * @deprcated since 3.6.0
```
```suggestion
	 * @deprecated since 3.5.0
```

StyleShit commented on includes/tracker.php:
```diff
@@ -162,9 +165,11 @@ public static function send_tracking_data( $override = false ) {
 	 */
 	public static function is_allow_track() {
-		return 'yes' === get_option( 'elementor_allow_tracking', 'no' );
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
+		return Site_Usage_Opt_In::is_on();
```
Is it 3.5.0 or 3.6.0?

```suggestion
		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track()', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
		return Site_Usage_Opt_In::is_on();
```

StyleShit commented on includes/user.php:
```diff
@@ -22,8 +24,14 @@ class User {
 	const ADMIN_NOTICES_KEY = 'elementor_admin_notices';
 
+	/**
+	 * @deprcated use User_Introduction
```
Also in other places.

```suggestion
	 * @deprecated 3.5.0 use `User_Introduction::get_key()` instead.
```

StyleShit commented on tests/phpunit/elementor/includes/settings/test-settings.php:
```diff
@@ -1,12 +1,17 @@
 
 	public function test_clear_css_cache_on_update_css_settings() {
+		// Register the option in order to listen to the changes.
+		Plugin::$instance->options->register( Site_CSS_Print_Method::class );
```
Isn't it registered by default?

https://github.com/elementor/elementor/blob/ae434047f20f115377d41b957e7c73040bf3962c/core/admin/admin.php#L687

matipojo replied:
It happens on the `admin.php`, but it seems that the test is running in another context

StyleShit commented on tests/phpunit/elementor/includes/test-tracker.php:
```diff
@@ -40,7 +42,7 @@ public function test_get_settings_advanced_usage() {
 		Plugin::$instance->icons_manager->register_admin_settings( Plugin::$instance->settings );
 
-		update_option( 'elementor_css_print_method', 'internal' );
+		Site_CSS_Print_Method::set( 'internal' );
```
Maybe that could be useful? Like `set_on()` / `set_off()`

```suggestion
		Site_CSS_Print_Method::set_internal();
```

StyleShit commented on core/options/site-option.php:
```diff
@@ -0,0 +1,28 @@
+	exit; // Exit if accessed directly
+}
+
+abstract class Site_Option extends Option_Base {
```
Tests?

StyleShit commented on core/options/option-base.php:
```diff
@@ -0,0 +1,98 @@
+	exit; // Exit if accessed directly
+}
+
+abstract class Option_Base {
```
Tests?

StyleShit reviewed (changes requested):
Great work! 😎

matipojo committed:
Review Fixes

matipojo committed:
Merge remote-tracking branch 'elementor/release/3.6.0' into tweak/options-manager

matipojo commented on core/admin/options/site-css-print-method.php:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	public static function get_default() {
+		return 'external';
```
Done

matipojo commented on core/admin/options/site-usage-notice.php:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	public static function set_off() {
+		return static::set( 1 );
```
It's saved as a string.

matipojo committed:
Fix PHP Unit test

matipojo committed:
Test Current User Option + Fixes

matipojo committed:
Replaced Current_User_Option with User_Option 

The User_Option for general use is redundant

matipojo committed:
Review Fixes

matipojo committed:
PHP Unit

matipojo committed:
Changed the `on_change` mechanism

matipojo committed:
Use an Error instead of an Exception for force implement a method

An Exception should be handled via `try... catch`

matipojo committed:
Cleanup

matipojo committed:
Validate parent of sub option

matipojo committed:
Test Option_Base

matipojo committed:
Changed the `on_change` mechanism

In order to listen to changes from wp-admin it needs to use the add/update option hooks

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
Remove unused setting

matipojo commented on includes/tracker.php:
```diff
@@ -170,9 +174,11 @@ public static function send_tracking_data( $override = false ) {
 	 */
 	public static function is_allow_track() {
-		return 'yes' === get_option( 'elementor_allow_tracking', 'no' );
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
```
```suggestion
		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.14.0', 'Site_Usage_Opt_In::is_on()' );
```

matipojo committed:
Update includes/tracker.php

matipojo commented on includes/user.php:
```diff
@@ -22,8 +24,14 @@ class User {
 	const ADMIN_NOTICES_KEY = 'elementor_admin_notices';
 
+	/**
+	 * @deprecated 3.6.0 use `User_Introduction::get_full_key()` instead.
```
```suggestion
	 * @deprecated 3.14.0 use `User_Introduction::get_full_key()` instead.
```

matipojo commented on includes/user.php:
```diff
@@ -22,8 +24,14 @@ class User {
 	const INTRODUCTION_KEY = 'elementor_introduction';
 
+	/**
+	 * @deprecated 3.6.0 use `Site_Beta::get_full_key()` instead.
```
```suggestion
	 * @deprecated 3.14.0 use `Site_Beta::get_full_key()` instead.
```

matipojo committed:
Update includes/user.php

matipojo committed:
Progress: Added Site_Font_Display & Site_Google_Fonts options

arielk commented:
Is this PR still relevant?

matipojo committed:
Progress

matipojo committed:
Merge branch 'main' into tweak/options-manager

rodolphebertozzo commented on core/admin/options/site-css-print-method.php:
```diff
@@ -16,6 +16,13 @@ public static function get_key() {
 
+	public static function get_options(){
+		return [
+			static::OPTION_EXTERNAL => __( 'External File', 'elementor' ),
+			static::OPTION_INTERNAL => __( 'Internal Embedding', 'elementor' ),
```
esc

rodolphebertozzo commented on core/options/boolean-option-base.php:
```diff
@@ -0,0 +1,35 @@
+
+	protected static function get_options() {
+		return [
+			static::OPTION_TRUE => __( 'True', 'elementor' ),
+			static::OPTION_FALSE => __( 'False', 'elementor' ),
```
esc

rodolphebertozzo commented on core/options/yes-no-option-base.php:
```diff
@@ -0,0 +1,33 @@
+
+	protected static function get_options() {
+		return [
+			static::OPTION_TRUE => __( 'Yes', 'elementor' ),
+			static::OPTION_FALSE => __( 'No', 'elementor' ),
```
esc

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
Fix: Namespaces

matipojo committed:
PHP Lint

matipojo committed:
Rename to Config + WP Option Base

matipojo committed:
New: WP options - Home, Blog Name, Blog Description and Site Icon

matipojo committed:
New: API Config - Remove Library & Remote Feed

matipojo committed:
Fix: PHPUnit failed

"Elementor\Core\Config\Config_Base::get must be implemented"

matipojo committed:
Fix: PHPUnit failed

"Error: Undefined class constant 'OPTION_NO'"

matipojo committed:
Fix: PHPUnit failed

"Class Elementor\Core\Admin\Config\Site_Usage_Opt_In cannot extend from trait"

matipojo committed:
Fix: PHPUnit failed

Error: Call to undefined method Elementor\Core\Admin\Config\Site_Google_Fonts::is_on()

matipojo committed:
Fix: Validation Methods

matipojo committed:
Fix: Test

matipojo committed:
Cannot override final method Site_Config_Base::setter

matipojo committed:
New: Maintenance Mode

matipojo committed:
Fix PHP

matipojo committed:
Merge branch 'main' into tweak/options-manager

matipojo committed:
Tweak: `validate` - Improved error message

matipojo committed:
Fix test

matipojo committed:
Fix tests

matipojo committed:
Fix tests

matipojo committed:
Tweak: `validate` - Improved error message

matipojo committed:
Fix: User config permissions

matipojo committed:
Fix tests: "the name is already in use"

matipojo committed:
Fix tests

matipojo committed:
Rename tests `options` => `config`

matipojo committed:
Fix: Set config during activate plugin is forbidden

matipojo committed:
Fix test failures

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
Escape HTML

matipojo committed:
Fix: `Delete` requires permissions

matipojo committed:
Fix: Test failures

matipojo committed:
Tweak: `set` is `final`

matipojo committed:
Fix PHP Parse error

matipojo committed:
Update core/admin/config/site-google-fonts.php

matipojo commented on core/admin/config/site-usage-last-send.php:
```diff
@@ -0,0 +1,26 @@
+	}
+
+	protected static function validate($value) {
+		return is_int($value);
```
```suggestion
		return is_int( $value );
```

matipojo committed:
Update core/admin/config/site-usage-last-send.php

matipojo commented on core/admin/config/wp-blog-description.php:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	protected static function validate($value) {
+		return is_scalar($value);
```
```suggestion
		return is_scalar( $value );
```

matipojo commented on core/admin/config/wp-blog-name.php:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	protected static function validate($value) {
+		return !empty($value) && is_string($value);
```
```suggestion
		return !empty( $value ) && is_string( $value );
```

matipojo commented on core/admin/config/wp-home.php:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	protected static function validate($value) {
+		return !empty($value) && is_string($value);
```
```suggestion
		return !empty( $value ) && is_string( $value );
```

matipojo commented on core/admin/config/wp-site-icon.php:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	protected static function validate($value) {
+		return is_scalar($value);
```
```suggestion
		return is_scalar( $value );
```

matipojo commented on core/api/config/site-remote-feed.php:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	protected static function validate($value) {
+		return is_array($value);
```
```suggestion
		return is_array( $value );
```

matipojo commented on core/api/config/site-remote-library.php:
```diff
@@ -0,0 +1,26 @@
+	}
+
+	protected static function validate($value) {
+		return is_array($value);
```
```suggestion
		return is_array( $value );
```

matipojo committed:
Apply suggestions from code review

matipojo committed:
PHP Unit on Windows

matipojo committed:
Fix tests

matipojo committed:
Lint and code arrangement

matipojo committed:
Fix `get_options` access level

matipojo committed:
PHPLint

matipojo committed:
Merge branch 'main' into tweak/options-manager

matipojo committed:
Fix: Allow set config on plugin activation

matipojo committed:
Fix: `register` should save class name

matipojo committed:
Fix: `is_admin` should be static to allow call before elementor/init

matipojo committed:
PHP Lint

matipojo committed:
PHP Lint

matipojo committed:
Merge branch 'main' into tweak/options-manager

matipojo commented on core/admin/admin.php:
```diff
@@ -839,6 +846,17 @@ public function init_beta_tester( $current_screen ) {
 	public function __construct() {
 		Plugin::$instance->init_common();
 
+		Plugin::$instance->config->register( Site_Beta::class );
+		Plugin::$instance->config->register( Site_CSS_Print_Method::class );
+		Plugin::$instance->config->register( Site_Font_Display::class );
+		Plugin::$instance->config->register( Site_Google_Fonts::class );
+		Plugin::$instance->config->register( Site_Usage_Last_Send::class );
+		Plugin::$instance->config->register( Site_Usage_Notice::class );
+		Plugin::$instance->config->register( Site_Usage_Opt_In::class );
+		Plugin::$instance->config->register( User_Beta_Tester::class );
+		Plugin::$instance->config->register( User_Introduction::class );
+		Plugin::$instance->config->register( WP_Blog_Name::class );
```
Remove

matipojo commented on core/config/config-base.php:
```diff
@@ -0,0 +1,143 @@
+	const VALUE_TRUE = true;
+	const VALUE_FALSE = false;
+
+	public static function get_options() {
```
Abstract, Move to a trait

matipojo commented on core/config/config-base.php:
```diff
@@ -0,0 +1,143 @@
+	final public static function set( $value ) {
+		$old_value = static::get();
+
+		if ( ! Manager::is_admin() ) {
```
```suggestion
		if ( ! Manager::is_backend() ) {
```

matipojo commented on core/config/config-base.php:
```diff
@@ -0,0 +1,143 @@
+			throw new \Error( 'Config can only be changed in the admin' );
+		}
+
+		// Avoid changing to a value that is not in the options.
+		if ( ! static::validate( $value ) ) {
+			throw new \Error( static::class . ': Invalid value: ' . var_export( $value, true ) );
+		}
```
Move to trait

matipojo commented on core/config/site-config-base.php:
```diff
@@ -0,0 +1,22 @@
+	/**
+	 * @return bool
+	 */
+	public static function should_autoload() {
+		throw new \Error( __METHOD__ . ' must be implemented' );
+	}
```
Remove

matipojo committed:
Added return types

matipojo committed:
Tweak: User_Introduction - Added `set_viewed` `is_viewed` methods

matipojo committed:
Tweak: Rename `get` => `get_value`

matipojo committed:
Internal: Use `abstract` in base classes

matipojo committed:
Fix tests and failings

matipojo committed:
Code Cleanup

matipojo committed:
Fix CLI permissions

matipojo committed:
Tweak: Rename `is_admin` => `is_backend`

matipojo committed:
Tweak: Rename `on_wp_change` => `on_external_change`

matipojo commented on core/config/config-base.php:
```diff
@@ -0,0 +1,143 @@
+		return static::deleter();
+	}
+
+	public static function get_prefix() {
+		return static::PREFIX;
+	}
+
+	public static function get_full_key() {
+		return static::get_prefix() . static::get_key();
+	}
```
Changed to an abstract method: `get_db_key`

matipojo committed:
Tweak: Rename `get_full_key` => `get_db_key`

matipojo committed:
Fix tests

matipojo committed:
Extract Array Trait

matipojo committed:
Fix tests

matipojo committed:
PHP Lint

matipojo committed:
PHP Lint

matipojo committed:
Refactor Tests

matipojo committed:
PHP Lint

matipojo committed:
New: Composer - `lint:fix` command

matipojo committed:
New: Composer - `coverage:test:html` command

matipojo committed:
gitignore coverage-report

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
New: Config - WP Page on Front and WP Show on Front

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
Merge remote-tracking branch 'elementor/main' into tweak/options-manager

matipojo committed:
wip
